### PR TITLE
DDP-7103 options groups for picklist study builder

### PR DESCRIPTION
**Pull request by aglushko-broad**
image

### CODE REVIEW

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-group-editor/picklist-group-editor.component.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            this.groupForm.enable();
+        }
+    }
+    @Input()
+    set notUniqueStableId(stableIdIsNotUnique: boolean) {
+        if (stableIdIsNotUnique) {
+            this.groupForm.get('stableId').setErrors({stableIdIsNotUnique: true});
+        } else {
+            this.groupForm.get('stableId').setErrors(null);
+        }
```
Comment (alextanlan) [inline ddp-workspace/projects/study-builder-ui/src/app/components/picklist-group-editor/picklist-group-editor.component.ts]: _**#minor**_

Can be shortened like 
```
const error = stableIdIsNotUnique ? {stableIdIsNotUnique: true} : null;
this.groupForm.get('stableId').setErrors(error);
```

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-option-editor/picklist-option-editor.component.ts [hunk 2]
```diff
@@ -29,6 +29,14 @@ export class PicklistOptionEditorComponent implements OnDestroy {
...
         this.optionForm.patchValue(optionValue);
     }
+    @Input()
+    set notUniqueStableId(stableIdIsNotUnique: boolean) {
+        if (stableIdIsNotUnique) {
+            this.optionForm.get('stableId').setErrors({stableIdIsNotUnique: true});
+        } else {
+            this.optionForm.get('stableId').setErrors(null);
+        }
+    }
```
Comment (alextanlan) [inline ddp-workspace/projects/study-builder-ui/src/app/components/picklist-option-editor/picklist-option-editor.component.ts:36]: see [comment above](https://github.com/broadinstitute/ddp-angular/pull/953/files#r756146752)

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts [hunk 3]
```diff
@@ -83,11 +83,18 @@ export class PicklistQuestionEditorComponent implements OnInit, OnDestroy {
...
     }
 
     private currentQuestion(): PicklistQuestionDef | null {
         return this.questionBlockSubject.getValue()?.question as PicklistQuestionDef;
     }
+
+    public allowCreateOptionsGroups(): boolean {
+        const selectMode = this.formGroup.get('selectMode').value;
+        const renderMode = this.formGroup.get('renderMode').value;
+        return selectMode === 'MULTIPLE' && renderMode === 'LIST' || renderMode === 'CHECKBOX_LIST' || renderMode === 'AUTOCOMPLETE';
```
Comment (alextanlan) [inline ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts:123]: _**#minor**_

can be shortened like
```
return selectMode === 'MULTIPLE' && ['LIST', 'CHECKBOX_LIST', 'AUTOCOMPLETE'].includes(renderMode);
```

**Review by alextanlan — DISMISSED** [review]

**Comment by MocanaAtBroad** [comment]
Looks good. Just a couple of little things I noticed.

**Comment by MocanaAtBroad** [comment]
I created two picklist questions, added groups and options to each.
But when I clicked one question and then the other (in the preview panel), the editor panel was getting updated (the question prompt indicated the selected question text). But the panel with the options and groups was stuck showing the options and groups for one question/was not being updated with the selection.

**Comment by MocanaAtBroad** [comment]
This I don't think is a bug, but did surprise me: When we create a picklist (checkbox list) with groups that have options, the groups are listed one on top of the other, but the options for each group are listed horizontally. I guess I was expecting that the default layout was to keep everything in vertical orientation. Could you just check that this indeed what we are doing by default (again, doubt this is a bug, but just something that surprised me)
Screen Shot 2021-11-29 at 6 11 14 PM
.

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts (hunk 3 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts:123]: not really, the logical operator will work like 
"(selectMode === 'MULTIPLE' && renderMode === 'LIST') || renderMode === 'CHECKBOX_LIST' || renderMode === 'AUTOCOMPLETE';"

**Comment by aglushko-broad** [comment]
> This I don't think is a bug, but did surprise me: When we create a picklist (checkbox list) with groups that have options, the groups are listed one on top of the other, but the options for each group are listed horizontally. I guess I was expecting that the default layout was to keep everything in vertical orientation. Could you just check that this indeed what we are doing by default (again, doubt this is a bug, but just something that surprised me) Screen Shot 2021-11-29 at 6 11 14 PM .

image
that's what I was able to find. In MPC project we put options into 2 columns. I'll copy all styles  for .ddp-picklist-group-container from MPC

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-group-editor/picklist-group-editor.component.ts (hunk 1 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/study-builder-ui/src/app/components/picklist-group-editor/picklist-group-editor.component.ts]: fixed

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-option-editor/picklist-option-editor.component.ts (hunk 2 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/study-builder-ui/src/app/components/picklist-option-editor/picklist-option-editor.component.ts:36]: fixed

**Comment by aglushko-broad** [comment]
> I created two picklist questions, added groups and options to each. But when I clicked one question and then the other (in the preview panel), the editor panel was getting updated (the question prompt indicated the selected question text). But the panel with the options and groups was stuck showing the options and groups for one question/was not being updated with the selection.

it was a bug, thanks, fixed!

**Review by MocanaAtBroad — APPROVED** [review]
Looks very good!

File: ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts (hunk 3 above)
Comment (alextanlan) [reply ddp-workspace/projects/study-builder-ui/src/app/components/picklist-question-editor/picklist-question-editor.component.ts:123]: My bad, agree.

**Pull request merged**